### PR TITLE
Top recoil hook cleaned, backport to CMSSW_13_2_X

### PR DESCRIPTION
**Pull request by vslokenb**
#### PR description:

Backport of recoilToTop. Pushed from 13_2_0_pre2.

#### PR validation:

Same as linked original PR below. Verified using `scram build code-check` and `scram build code-format`. Seem to be a couple of minor indentation changes in lines unrelated to the actual TopRecoilHook in Pythia8Hadronizer.cc as a result of the code-format tool.  All differences between original PR are to account for CMSSW_13_2_X formatting needs.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Original PR found [here](https://github.com/cms-sw/cmssw/pull/42180). Backport requested by maintainers.

### CODE REVIEW

**Comment by perrotta** [comment]
backport of #42180

**Comment by menglu21** [comment]
please test

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+      // Histogram weight ratio.
+      wtCorr->fill(wtT / wtW);
+
+      // List relevant properties.
+      if (doList) {
+        partonSystemsPtr->list();
+        event.list();
+      }
+      //std::cout << "BEGIN TOP TEST: "  << (wtT / wtW) << " | END TOP TEST" << endl;
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:120]: Please remove
```suggestion
```

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    TopRecoilHook(bool doTopRecoilIn = true, bool useOldDipoleIn = false, bool doListIn = false) {
+      doTopRecoil = doTopRecoilIn;
+      useOldDipole = useOldDipoleIn;
+      doList = doListIn;
+      // Constructor also creates some histograms for analysis inside User Hook.
+      wtCorr = new Hist("corrective weight", 100, 0., 2.);
+    }
+
+    // Destructor prints histogram.
+     ~TopRecoilHook() override { delete wtCorr; }
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:36]: Please align
```suggestion
    ~TopRecoilHook() override { delete wtCorr; }
```

**Review by perrotta — COMMENTED** [review]
This being based on the rather old CMSSW_3_2_0_pre2, it was a bit complicate to compare with the same PR merged in master.
At the end I managed to do it.: therefore, even if a rebase on the more recent 13_2_X IB could help, it is not mandatory, provided you apply the two small adjustments listed here needed to synchronize with the master version of the same PR.
After that (and the @cms-sw/generators-l2 signature, of course) this PR can be merged

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 3]
```diff
@@ -29,7 +30,6 @@ using namespace Pythia8;
 // Emission Veto Hooks
 //
 #include "Pythia8Plugins/PowhegHooks.h"
-#include "Pythia8Plugins/PowhegHooksVincia.h"
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:32]: This PR backports in 13_2_X also https://github.com/cms-sw/cmssw/pull/42020
However, that PR is supposed to go with https://github.com/cms-sw/cmsdist/pull/8552, which as far as I can see is not backported in 13_2.
Please backport in 13_2 only what is needed for this PR. Updates of Pythia in closed release cycles are delicate and should be either avoided or dealt with by @cms-sw/generators-l2 in a separate and dedicated PR

**Comment by perrotta** [comment]
hold
(see https://github.com/cms-sw/cmssw/pull/42332#pullrequestreview-1591107324)

**Comment by menglu21** [comment]
@vslokenb kind ping

**Comment by vslokenb** [comment]
Thanks for the reminder, hopefully this fixes the issue!

**Comment by perrotta** [comment]
unhold
(now this PR truly reflects what is in the backported PRs #42180 and #42313)

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 4]
```diff
@@ -19,6 +19,7 @@ using namespace Pythia8;
 
 #include "GeneratorInterface/Pythia8Interface/plugins/ReweightUserHooks.h"
```
Comment (menglu21) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:20]: in https://github.com/cms-sw/cmssw/pull/42180/, it's using "#include "ReweightUserHooks.h""

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 5]
```diff
@@ -19,6 +19,7 @@ using namespace Pythia8;
...
 #include "GeneratorInterface/Pythia8Interface/interface/CustomHook.h"
+#include "GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h"
```
Comment (menglu21) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:22]: in https://github.com/cms-sw/cmssw/pull/42180, it's using "#include "TopRecoilHook.h"

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    TopRecoilHook(bool doTopRecoilIn = true, bool useOldDipoleIn = false, bool doListIn = false) {
+      doTopRecoil = doTopRecoilIn;
+      useOldDipole = useOldDipoleIn;
+      doList = doListIn;
+      // Constructor also creates some histograms for analysis inside User Hook.
+      wtCorr = new Hist("corrective weight", 100, 0., 2.);
+    }
+
+    // Destructor prints histogram.
+    ~TopRecoilHook() override { delete wtCorr; }
```
Comment (menglu21) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:36]: in https://github.com/cms-sw/cmssw/pull/42180, it's:
```
 ~TopRecoilHook() override {
      if (doTopRecoil)
        ;
      delete wtCorr;
    }
```

**Comment by menglu21** [comment]
please test

**Comment by menglu21** [comment]
@cmsbuild please test

**Comment by menglu21** [comment]
@perrotta is the error related to this PR

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+      // Constructor also creates some histograms for analysis inside User Hook.
+      wtCorr = new Hist("corrective weight", 100, 0., 2.);
+    }
+
+    // Destructor prints histogram.
+     ~TopRecoilHook() override {
+      if (doTopRecoil)
+        ;
+      delete wtCorr;
+    }
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:40]:
```suggestion
    ~TopRecoilHook() override { delete wtCorr; }
```

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h (hunk 6 above)
Comment (perrotta) [reply GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h]: Please don't revert this fix, which was merged in master with #42313 !
(Se also my comment https://github.com/cms-sw/cmssw/pull/42332#issuecomment-1755380830 above, when I unheld this PR)

**Comment by menglu21** [comment]
@cmsbuild please test

**Comment by menglu21** [comment]
+1

**Comment by rappoccio** [comment]
@perrotta are you okay with the PR now? Have your comments been addressed?

**Comment by perrotta** [comment]
> @perrotta are you okay with the PR now? Have your comments been addressed?

Ciao Sal. Yes, now this PR correctly implements the fix that was merged in master with https://github.com/cms-sw/cmssw/pull/42313

**Comment by antoniovilela** [comment]
+1

**Pull request merged**